src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs

```diff
@@ -105,43 +105,48 @@ internal SdkDirectoryWorkloadManifestProvider(string sdkRootPath, string sdkVers
             RefreshWorkloadManifests();
         }
 
-        public void RefreshWorkloadManifests()
+        public void RefreshWorkloadManifests(bool error = false)
         {
+            _workloadSet = null;
+            _manifestsFromInstallState = null;
+            _installStateFilePath = null;
             var availableWorkloadSets = GetAvailableWorkloadSets();
 
             if (_workloadSetVersionFromConstructor != null)
             {
-                if (!availableWorkloadSets.TryGetValue(_workloadSetVersionFromConstructor, out _workloadSet))
+                if (!availableWorkloadSets.TryGetValue(_workloadSetVersionFromConstructor, out _workloadSet) && error)
```

Comment:
@dsplaisted,

This is the part that I was poking at most recently, and I mentioned I'm not totally satisfied with its current state, but it sounded like it should mostly not error; this is just a bit more generous than that...

Reply:
After some thought, here's how I think it should work:

The scenario where it currently fails but we don't want it to fail is when we are going to install the requested workload set version.  In that case, I think that most of the manifest provider and workload resolver functionality probably isn't used until we've downloaded and installed the requested workload set version and corresponding manifests, and then we call RefreshWorkloadManifests.

So what we should do is to delay throwing the error until the `GetManifests` method is called.  Then we'll never actually end up throwing it in the case where we immediately install the needed workload set.

It looks like `WorkloadResolver` loads the manifests during its construction, so we would also need to modify that class to enable it to delay the initialization.  It's possible that might cascade into other code that uses it too.

We may want to only delay the initialization when we are running commands such as install or update and possibly only when there is a global.json applicable.  That would be to avoid changing existing behavior around when exceptions are thrown and what the stack trace is.  It may not really matter though.